### ISSUE TITLE
ImmutableTable.get results in copyOf with actual copy being invoked

### ISSUE THREAD

**Issue by victornoel** [google/guava#3467]
I faced a strange performance situation with `ImmutableTable`.

Here is a stacktrace of what's happening (it's a profiling trace from [async-profiler](https://github.com/jvm-profiling-tools/async-profiler)):
![image](https://user-images.githubusercontent.com/160975/57542430-f7e47300-7351-11e9-83a6-10f8c42191b9.png)

Basically, it seems that calling `ImmutableTable.get()` results in `ImmutableMap.copyOf()` being called, and even though there is a comment in `rowMap()` saying that `copyOf` is called only for casting (https://github.com/google/guava/blob/master/guava/src/com/google/common/collect/SparseImmutableTable.java#L97), in practice, there is an actual copy via `toArray()`.

I believe this is not meant to happen, is it?

**Comment by victornoel** [comment]
For the record, the `ImmutableTable` was created using `ImmutableTable.toImmutableTable`

**Comment by samukce** [comment]
Hello, I was looking at this and I could understand why this is happening for `get` method. Basically, the classes that:
1. Inherits from AbstractTable;
2. It doesn't override the `V get` method;
3. It has in `rowmap()` implementation the use of `copyOf` or `ImmutableMap.of`

So based on those criteria I could find those classes that have this same behavior of creating news maps internally:
- SparseImmutableTable
- SingletonImmutableTable
- DenseImmutableTable

So, trying to summarize my founds: every use of rowMap()(+columnMap()) method in AbstractTable class it's a potential point of creating extra objects that it seems not necessary for such methods implementations such as: `containsRow, containsColumn, containsValue, contains, get` for ImmutableTable implementations;

My initial thought about it is that the `rowMap` and `columnMap` method it has been burdened with responsibilities: 
1. to provide external data (which can be mutable or immutable data) 
2. to provide support internally to access the data.

That said I can see this possible approach: To have an internal way to access the `rowMap/columnMap` in a protected class scope from ImmutableTable class so we could have those methods been implemented by this one, avoiding create unnecessary objects; 

Something like 
```java
public abstract class ImmutableTable<R, C, V> extends AbstractTable<R, C, V> {
....
  protected abstract ImmutableMap<R, ImmutableMap<C, V>> immutableRowMap();

  @Override
  public boolean containsRow(@Nullable Object rowKey) {
    return Maps.safeContainsKey(immutableRowMap(), rowKey);
  }
...
}

final class DenseImmutableTable<R, C, V> extends RegularImmutableTable<R, C, V> {
  @Override
  protected ImmutableMap<R, ImmutableMap<C, V>> immutableRowMap() {
    return this.rowMap;
  }
```

Thoughts @kluever / @ronshapiro ?

**Comment by samukce** [comment]
I realized that the `ImmutableMap.copyOf` creation of new objects will not happen if: the map is it's an ImmutableMap and Is not SortedMap and is not PartialView; Also, if It is an EnumMap it will generate a copy and for the other map types will generate a copy too;

```java
public static <K, V> ImmutableMap<K, V> copyOf(Map<? extends K, ? extends V> map) {
    if ((map instanceof ImmutableMap) && !(map instanceof SortedMap)) {
      @SuppressWarnings("unchecked") // safe since map is not writable
      ImmutableMap<K, V> kvMap = (ImmutableMap<K, V>) map;
      if (!kvMap.isPartialView()) {
        return kvMap;
      }
    } else if (map instanceof EnumMap) {
      @SuppressWarnings("unchecked") // safe since map is not writable
      ImmutableMap<K, V> kvMap = (ImmutableMap<K, V>) copyOfEnumMap((EnumMap<?, ?>) map);
      return kvMap;
    }
    return copyOf(map.entrySet());
  }
```

https://github.com/google/guava/blob/master/guava/src/com/google/common/collect/ImmutableMap.java#L426-L439

**Comment by samukce** [comment]
In the end, I could not reproduce my hypnotizes either using the regular builder or the immutable collection (`ImmutableTable.toImmutableTable`).   😢  For both cases when we hit the `copyOf` method to `rowMap()` method through the `get` method, this map is never recreated returning the own object [here](https://github.com/google/guava/blob/master/guava/src/com/google/common/collect/ImmutableMap.java#L431) Probably I'm missing something here...  any help is welcome. 🤞

**Comment by cpovirk** [comment]
This does seem very strange, and it pops into my head from time to time.

The only even remotely plausible theory I have is that the `ImmutableTable` is an instance of the type from one classloader and the `ImmutableMap` call is using the type from another.

**Comment by samukce** [comment]
I have tried to create a test to reproduce the issue but I could not. So created one just to guarantee no new object is created per call. Do you have any idea how to create this test for the cases you mentioned @cpovirk ? 

https://github.com/google/guava/blob/51a76c04892de73578a2831897fb53fff9963e86/guava-tests/test/com/google/common/collect/ImmutableTableTest.java#L345-L363

**Comment by cpovirk** [comment]
For what it's worth, I'm confident that the code _would_ behave that way _if_ we were seeing that kind of interaction between multiple classloaders. What I'm not at _all_ confident in is that such an interaction was happening for the original poster.

If you're curious, it's definitely possible to test such things, but it's definitely a pain :) You can see some examples custom classloaders in our tests, such as in this test: https://github.com/google/guava/blob/081c486173032e6096c912e3c297c1d74eddcf93/guava-tests/test/com/google/common/util/concurrent/AbstractFutureInnocuousThreadTest.java#L48

**Comment by cpovirk** [comment]
Since I've been looking a little at other unnecessary allocation in `common.collect`, I just looked at our fleet-wide profiling data to see if we had any detectable allocations happening anywhere inside `*ImmutableTable.rowMap()`. The only allocations I found were for `SingletonImmutableTable.rowMap()`, which is expected: It goes through `ImmutableMap.of` (not `copyOf`), and its allocations are negligible.

Given that, and given that it's been a few years, I'm going to close this. I don't know if any of [the async-profiler limitations](https://github.com/async-profiler/async-profiler?tab=readme-ov-file#restrictionslimitations) might apply here: Maybe "in some rare cases, the captured Java stack might not match the captured native (user+kernel) stack" means that an allocation could be getting blamed on a subsequent `get` call, for example? I wouldn't feel comfortable dismissing this report based just on that, but realistically, there's probably nothing we can do but assume that we're seeing _something_ like that.
